README.md

```diff
@@ -1,28 +1,176 @@
+  could publish the API response with a routing key of `user.commits.fetched`.
+  This documents what already happened and makes no assumptions about what
+  happens next.
+
+## Message validation
+
+Velveteen can validate incoming messages with a JSON Schema.
+
+```ruby
+Velveteen::Config.schema_directory = "app/schemas"
+
+class FetchUserCommits < Velveteen::Worker
+  ...
+  self.message_schema = "fetch_user_commits.json"
+
+  def perform
+    # do something with the GitHub API
+  end
+end
 ```
 
-## Supported Ruby versions
+_app/schema/fetch_user_commits.json_
+```json
+{
+  "type": "object",
+  "required": ["username"],
+  "properties": {
+    "username": {"type": "string"}
+  }
+}
+```
 
-TODO: Figure this out
+When a message is added to this worker's queue and it does not match the schema,
+a `Velveteen::InvalidMessage` exception be raised and handled by the configured
+error handler. The worker's `perform` method will not be invoked.
+
+## Rate limiting
+
+Limiting workers to processing a certain number of messages per minute is
+supported. This is implemented by periodically publishing tokens to a dedicated
+queue (with a max size of 1), where the worker must successfully take a token
+before it begins.
+
+To produce tokens, Velveteen needs the name of the queue to publish to and the
+number of messages per minute. For example, [GitHub's API limit] is 5000
+requests per hour (~83.3 per minute). With Velveteen, that looks like:
+
+[GitHub's API limit]: https://developer.github.com/v3/#rate-limiting
+
+```shell
+velveteen rate-limit github_tokens 83
+```
+
+Limiting the worker can be enabled by specifying the rate limit queue:
+
+```ruby
+class FetchCommits < Velveteen::Worker
+  ...
+  self.rate_limit_queue = "github_tokens"
+
+  def perform
+    # do something with the GitHub API
+  end
+end
+```
+
+Multiple workers can share the same token queue, which can be helpful when API
+calls to an external service are spread among many workers and are all governed
+by the same rate limit.
+
+The provided implementation does not support other techniques, such as bursting,
+throttling, or quotas.
+
+## Error handling
+
+### Message rejection
+
+The default error handler will reject the message without requeueing it and log
+the error.
+
+### Retry with exponential backoff
+
+An error handler that supports exponential backoff retries is included. It can
+be enabled with:
+
+```ruby
+Velveteen::Config.error_handler = Velveteen::ErrorHandlers::ExponentialBackoff
+```
+
+## Process model
+
+In short, there isn't one. Velveteen supports running a single worker that
+consumes messages from a single queue. Process management is an external
+concern left to application developers.
+
+The approach in mind while developing/extracting Velveteen was to have a single
+entry for each worker in a Procfile, hosted on Heroku, using hobby tier dynos.
+This allows for individually scaling workers as needed, without the application
```

Comment:
I forgot hobby tier doesn't support horizontal scaling - you have to upgrade to standard-1x for that - https://devcenter.heroku.com/articles/dyno-types#dyno-feature-support

On the project Velveteen was extracted from, we never ran into the need to scale out the workers because the rate limits of external APIs were the bottleneck of every path and a single worker could easily keep up with them.

For now, it probably makes sense to reword or remove this part about scaling and rethink this overall setup when a scaling problem arises.

Reply:
Not sure if it's worth pointing out in the readme, but for brevity here, there is also a limit of 100 total dynos across all the processes for an app, so there is a hard maximum on the number of workers you can run with this approach.